src/main/java/macrobase/analysis/outlier/MovingAverage.java

```diff
@@ -0,0 +1,88 @@
+package macrobase.analysis.outlier;
+
+import java.util.Iterator;
+import java.util.Queue;
+import java.util.concurrent.ArrayBlockingQueue;
+
+import org.apache.commons.math3.linear.ArrayRealVector;
+import org.apache.commons.math3.linear.RealVector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import macrobase.datamodel.Datum;
+
+/*
+ * Simple moving average over past `timeSeriesWindowSize` data points.
+ */
+public class MovingAverage extends TimeSeriesOutlierDetector {
+    private static final Logger log = LoggerFactory.getLogger(MovingAverage.class);
+    int prevTime;
+    Queue<Integer> weights;
+    int weightTotal;
+    
+    public MovingAverage(int timeSeriesWindowSize) {
+        super(timeSeriesWindowSize);
+    }
+
+    private RealVector windowSum;
+
+    @Override
+    public void initWindow(Queue<Datum> data) {
+        windowSum = null;
+    }
+
+    @Override
+    public void trainWindow(Queue<Datum> data, Datum newDatum, Datum oldDatum) {
+        /* Empty */
+    }
+
+    @Override
+    public double scoreWindow(Queue<Datum> data, Datum newDatum, Datum oldDatum) {
```

Comment:
There's a distinction to be made between the interface for algorithm writers and for algorithm users. For algorithm writers, I think it's generally easier to think in terms of tuple-at-a-time, and this interface is designed for them. For consumers of the algorithm's scoring, it may be easier to use the panes idea, but that can be a different interface/class that calls the TimeSeriesOutlierDetector interface inside.